packages/z3/z3.4.8.5/files/4468.patch

```diff
@@ -0,0 +1,267 @@
+shared library that could be loaded in runtime use
+`Dynlink.loadfile` (which used dlopen on posix machines underneath the
+hood).
+
+- libz3ml.a is the archived machine code for `z3native_stubs.c`, which
+is made by ocamlmklib: `ar rcs api/ml/libz3ml.a
+api/ml/z3native_stubs.o` it is needed to build statically linked
+binaries and libraries that use z3 bindings.
+
+- dllz3ml.so is the shared object that contains `z3native_stubs.o` as
+well as correct ldd entries for C++ and Z3 libraries to enable proper
+static and dynamic linking. The file is built with ocamlmklib on posix
+systems as
+```
+gcc -shared -o api/ml/dllz3ml.so api/ml/z3native_stubs.o -L. -lz3-static -lstdc++
+```
+
+It is used by `ocaml`, `ocamlrun`, and `ocamlc` to link z3 and c++
+code into the OCaml runtime and enables usage of z3 bindings in
+non-custom runtimes (default runtimes).
+
+The `dllz3ml.so` is usually installed in the stubs library in opam
+installation (`$(opam config var lib)/stublibs`), it is done
+automatically by `ocamlfind` so no special treatment is needed.
+
+Technical Details
+=================
+
+The patch itself is rather small. First of all, we have to use
+`-l<lib>` instead of `-cclib -l<lib>` in ocamlmklib since the latter
+will pass the options only to the ocaml{c,opt} linker and will not
+use the passed libraries when shared and non-shared versions of the
+bindings are built (libz3ml.a and dllz3ml.so). They were both missing
+either z3 code itself and ldd entries for stdc++ (and z3 if built not
+in --staticlib mode).
+
+Having stdc++ entry streamlines the compilation process and makes
+dynamic loading more resistant to the inclusion order.
+
+Finally, we had to add `-L.` to make sure that the built artifacts are
+correctly found by gcc.
+
+I specifically left the cygwin part of the code intact as I have no
+idea what the original author meant by this, neither do I use or
+tested this patch in the cygwin or mingw environemt. I think that this
+code is rather outdated and shouldn't really work. E.g., in the
+--staticlib mode adding z3linkdep (which is libz3-static.a) as an
+argument to `ocamlmklib` will yield the following broken archive
+```
+ar rcs api/ml/libz3ml.a libz3-static.a api/ml/z3native_stubs.o
+```
+and it is not allowed (or supported) to have .a in archives (though it
+doesn't really hurt as most of the systems will just ignore it).
+
+But otherwise, cygwin, mingw shall behave as they did (the only change
+that affects them is `-L.` which I believe should be benign).
+---
+ scripts/mk_util.py | 6 +++---
+ 1 file changed, 3 insertions(+), 3 deletions(-)
+
+diff --git a/scripts/mk_util.py b/scripts/mk_util.py
+index 1a95ee20c3..156be9ca2f 100644
+--- a/scripts/mk_util.py
++++ b/scripts/mk_util.py
+@@ -1957,7 +1957,7 @@ def mk_makefile(self, out):
+ 
+             OCAMLMKLIB = 'ocamlmklib'
+ 
+-            LIBZ3 = '-cclib -l' + z3link
++            LIBZ3 = '-l' + z3link + ' -lstdc++'
```

Comment:
@floriangru, we need to add the missing libgomp here, i.e., `-lgomp`.

Reply:
@catalin-hritcu do you know about this issue?

Reply:
@mseri: Not really, sorry. I'm one of the developers of F* who originally encouraged @floriangru to submit this pull request, because for us exactly this version of Z3 is a dependency. Neither me nor @floriangru knows much about Z3 packing, but we were hoping it would be easy :)

Reply:
Ok, thanks for letting us know. Let's try to unlock this PR.

@ivg, since you seem to know what is missing, would you have time to take over the PR and add the missing `-lgomp` where needed?

Reply:
> Neither me nor @floriangru knows much about Z3 packing, but we were hoping it would be easy :)

Normally it is, but it depends very much on how complex is the package and its set of external dependencies.

Reply:
@mseri, sure, I will take a loot into it.

Reply:
Thank you very much